pinot-core/src/main/java/org/apache/pinot/core/transport/QueryRouter.java

```diff
@@ -130,6 +146,10 @@ public AsyncQueryResponse submitQuery(long requestId, String rawTableName,
         serverChannels.sendRequest(rawTableName, asyncQueryResponse, serverRoutingInstance, entry.getValue(),
             timeoutMs);
         asyncQueryResponse.markRequestSubmitted(serverRoutingInstance);
+
+        if (limitClausePresent && --limit == 0) {
```

Comment:
I am guessing we also want to make changes to send cancellation request after early termination ?

Reply:
query cancellation work can potentially be leveraged for that.